diff --git a/package/boot/kexec-tools/Makefile b/package/boot/kexec-tools/Makefile
index 7f06ec7d64..dc9464e383 100644
--- a/package/boot/kexec-tools/Makefile
+++ b/package/boot/kexec-tools/Makefile
@@ -43,7 +43,7 @@ define Package/kexec
   $(call Package/kexec-tools/Default)
   TITLE:=Kernel boots kernel
   DEPENDS:=\
-       @(armeb||arm||i386||x86_64||powerpc64||mipsel||mips) \
+       @(armeb||arm||i386||x86_64||powerpc64||mipsel||mips||aarch64) \
        +KEXEC_ZLIB:zlib +KEXEC_LZMA:liblzma @KERNEL_KEXEC
 endef

@@ -54,7 +54,7 @@ endef
 define Package/kdump
   $(call Package/kexec-tools/Default)
   TITLE:=Kernel crash analysis
-  DEPENDS:=+kexec @(i386||x86_64||arm) @KERNEL_CRASH_DUMP
+  DEPENDS:=+kexec @(i386||x86_64||arm||aarch64) @KERNEL_CRASH_DUMP
 endef

 define Package/kdump/description