diff --git a/package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx b/package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx
--- package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx
+++ package/boot/uboot-tools/uboot-envtools/files/qualcommax_ipq60xx
@@ -24,14 +24,18 @@
 qihoo,v6|\
 redmi,ax5|\
 xiaomi,ax1800|\
 glinet,gl-ax1800|\
-glinet,gl-axt1800|\
 netgear,wax214)
 	idx="$(find_mtd_index 0:appsblenv)"
 	[ -n "$idx" ] && \
 		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x40000" "0x20000"
 	;;
+glinet,gl-axt1800)
+	idx="$(find_mtd_index 0:appsblenv)"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x40000" "0x40000"
+	;;
 jdcloud,re-cs-07|\
 redmi,ax5-jdcloud|\
 jdcloud,re-ss-01|\
 jdcloud,re-cs-02)
