diff --git a/arch/arm64/boot/dts/qcom/ipq6018.dtsi b/arch/arm64/boot/dts/qcom/ipq6018.dtsi
--- a/arch/arm64/boot/dts/qcom/ipq6018.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq6018.dtsi
@@ -27,6 +27,18 @@
 			clock-frequency = <24000000>;
 			#clock-cells = <0>;
 		};
+		
+		bias_pll_cc_clk: bias-pll-cc-clk {
+			compatible = "fixed-clock";
+			clock-frequency = <300000000>;
+			#clock-cells = <0>;
+		};
+
+		bias_pll_nss_noc_clk: bias-pll-nss-noc-clk {
+			compatible = "fixed-clock";
+			clock-frequency = <416500000>;
+			#clock-cells = <0>;
+		};
 	};
 
 	cpus: cpus {
@@ -247,8 +259,14 @@
 		gcc: gcc@1800000 {
 			compatible = "qcom,gcc-ipq6018";
 			reg = <0x0 0x01800000 0x0 0x80000>;
-			clocks = <&xo>, <&sleep_clk>;
-			clock-names = "xo", "sleep_clk";
+			clocks = <&xo>,
+				 <&sleep_clk>,
+				 <&bias_pll_cc_clk>,
+				 <&bias_pll_nss_noc_clk>;
+			clock-names = "xo",
+				      "sleep_clk",
+				      "bias_pll_cc_clk",
+				      "bias_pll_nss_noc_clk";
 			#clock-cells = <1>;
 			#reset-cells = <1>;
 		};
@@ -373,6 +391,124 @@
 			status = "disabled";
 		};
 
+		sdcc1_ice: sdcc1ice@7808000 {
+			compatible = "qcom,ice";
+			reg = <0x7808000 0x2000>;
+			interrupts = <GIC_SPI 312 IRQ_TYPE_NONE>;
+			qcom,msm-bus,vectors-KBps =
+				<78 512 0 0>,    /* No vote */
+				<78 512 1000 0>; /* Max. bandwidth */
+			qcom,bus-vector-names = "MIN", "MAX";
+			qcom,instance-type = "sdcc";
+		};
+
+		sdhc_1: sdhci@7804000 {
+			compatible = "qcom,sdhci-msm-v5";
+			reg = <0x7804000 0x1000>, <0x7805000 0x1000>;
+			reg-names = "hc_mem", "cmdq_mem";
+
+			interrupts = <GIC_SPI 123 IRQ_TYPE_NONE>,
+					<GIC_SPI 138 IRQ_TYPE_NONE>;
+			interrupt-names = "hc_irq", "pwr_irq";
+
+			qcom,bus-width = <8>;
+			sdhc-msm-crypto = <&sdcc1_ice>;
+			qcom,max_clk = <192000000>;
+
+			qcom,dedicated-io = <1>;
+
+			/* device core power supply */
+			qcom,vdd-voltage-level = <2900000 2900000>;
+			qcom,vdd-current-level = <200 570000>;
+
+			/* device communication power supply */
+			qcom,vdd-io-lpm-sup;
+			qcom,vdd-io-voltage-level = <1800000 1800000>;
+			qcom,vdd-io-current-level = <200 325000>;
+			qcom,vdd-io-always-on;
+
+			qcom,cpu-dma-latency-us = <701>;
+			qcom,msm-bus,name = "sdhc1";
+			qcom,msm-bus,num-cases = <9>;
+			qcom,msm-bus,num-paths = <1>;
+			qcom,msm-bus,vectors-KBps = <78 512 0 0>, /* No vote */
+				<78 512 1046 3200>,    /* 400 KB/s*/
+				<78 512 52286 160000>, /* 20 MB/s */
+				<78 512 65360 200000>, /* 25 MB/s */
+				<78 512 130718 400000>, /* 50 MB/s */
+				<78 512 261438 800000>, /* 100 MB/s */
+				<78 512 261438 800000>, /* 200 MB/s */
+				<78 512 261438 800000>, /* 400 MB/s */
+				<78 512 1338562 4096000>; /* Max. bandwidth */
+			qcom,bus-bw-vectors-bps = <0 400000 20000000 25000000 \
+						50000000 100000000 200000000 \
+						400000000 4294967295>;
+
+			clocks = <&gcc GCC_SDCC1_AHB_CLK>,
+				 <&gcc GCC_SDCC1_APPS_CLK>,
+				 <&gcc GCC_SDCC1_ICE_CORE_CLK>;
+			clock-names = "iface_clk", "core_clk", "ice_core_clk";
+			qcom,ice-clk-rates = <160000000 308570000>;
+			qcom,large-address-bus;
+			qcom,disable-aggressive-pm;
+			qcom,clk-rates = <400000 25000000 50000000 100000000 \
+			 192000000>;
+			qcom,bus-speed-mode = "HS200_1p8v", "DDR_1p8v";
+			qcom,nonremovable;
+
+			status = "disabled";
+		};
+
+		sdhc_2: sdhci_sd@7804000 {
+			compatible = "qcom,sdhci-msm-v5";
+			reg = <0x7804000 0x1000>;
+			reg-names = "hc_mem";
+
+			interrupts = <GIC_SPI 123 IRQ_TYPE_NONE>,
+					<GIC_SPI 138 IRQ_TYPE_NONE>;
+			interrupt-names = "hc_irq", "pwr_irq";
+
+			qcom,max_clk = <192000000>;
+			qcom,bus-width = <4>;
+			qcom,dedicated-io = <1>;
+			/* device core power supply */
+			qcom,vdd-voltage-level = <2950000 2950000>;
+			qcom,vdd-current-level = <15000 400000>;
+
+			/* device communication power supply */
+			qcom,vdd-io-lpm-sup;
+			qcom,vdd-io-voltage-level = <1800000 2950000>;
+			qcom,vdd-io-current-level = <200 22000>;
+			qcom,vdd-io-always-on;
+
+			qcom,cpu-dma-latency-us = <701>;
+			qcom,msm-bus,name = "sdhc2";
+			qcom,msm-bus,num-cases = <8>;
+			qcom,msm-bus,num-paths = <1>;
+			qcom,msm-bus,vectors-KBps = <81 512 0 0>, /* No vote */
+				<81 512 1046 3200>,    /* 400 KB/s*/
+				<81 512 52286 160000>, /* 20 MB/s */
+				<81 512 65360 200000>, /* 25 MB/s */
+				<81 512 130718 400000>, /* 50 MB/s */
+				<81 512 261438 800000>, /* 100 MB/s */
+				<81 512 261438 800000>, /* 200 MB/s */
+				<81 512 1338562 4096000>; /* Max. bandwidth */
+			qcom,bus-bw-vectors-bps = <0 400000 20000000 25000000 \
+						50000000 100000000 200000000 \
+						4294967295>;
+
+			clocks = <&gcc GCC_SDCC1_AHB_CLK>,
+				 <&gcc GCC_SDCC1_APPS_CLK>;
+			clock-names = "iface_clk", "core_clk";
+			qcom,large-address-bus;
+			qcom,disable-aggressive-pm;
+			qcom,clk-rates = <400000 25000000 50000000 100000000 \
+			 192000000>;
+			qcom,bus-speed-mode = "HS200_1p8v", "DDR_1p8v";
+
+			status = "disabled";
+		};
+
 		intc: interrupt-controller@b000000 {
 			compatible = "qcom,msm-qgic2";
 			interrupt-controller;
@@ -600,8 +736,24 @@
 				      "wcss_reset",
 				      "wcss_q6_reset";
 
-			clocks = <&gcc GCC_PRNG_AHB_CLK>;
-			clock-names = "prng";
+			clocks = <&gcc GCC_PRNG_AHB_CLK>,
+					<&gcc GCC_SYS_NOC_WCSS_AHB_CLK>,
+					<&gcc GCC_Q6SS_ATBM_CLK>,
+					<&gcc GCC_Q6SS_PCLKDBG_CLK>,
+					<&gcc GCC_Q6_TSCTR_1TO2_CLK>;
+			clock-names = "prng",
+					"gcc_sys_noc_wcss_ahb_clk",
+					"gcc_q6ss_atbm_clk",
+					"gcc_q6ss_pclkdbg_clk",
+					"gcc_q6_tsctr_1to2_clk";
+			assigned-clocks = <&gcc GCC_SYS_NOC_WCSS_AHB_CLK>,
+					  <&gcc GCC_Q6SS_PCLKDBG_CLK>,
+					<&gcc GCC_Q6_TSCTR_1TO2_CLK>,
+					<&gcc GCC_Q6SS_ATBM_CLK>;
+			assigned-clock-rates = <133333333>,
+						<600000000>,
+						<600000000>,
+						<240000000>;
 
 			qcom,halt-regs = <&tcsr 0x18000 0x1b000 0xe000>;
 
@@ -623,6 +775,247 @@
 			};
 		};
 
+		mdio: mdio@90000 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			compatible = "qcom,ipq6018-mdio", "qcom,ipq4019-mdio";
+			reg = <0x0 0x90000 0x0 0x64>;
+			clocks = <&gcc GCC_MDIO_AHB_CLK>;
+			clock-names = "gcc_mdio_ahb_clk";
+			status = "disabled";
+		};
+
+		ssphy_0: ssphy@78000 {
+			compatible = "qcom,usb-ssphy-qmp";
+			reg = <0x78000 0x45c>,
+				<0x0193f244 0x4>,
+				<0x08af8800 0x100>,
+				<0x7e000 0x18>;
+			reg-names = "qmp_phy_base",
+				    "vls_clamp_reg",
+				    "qscratch_base",
+				    "ahb2phy_base";
+			qcom,qmp-phy-init-seq = <0xac 0x14 0x1a 0x00
+						0x34 0x08 0x08 0x00
+						0x174 0x30 0x30 0x00
+						0x3c 0x06 0x06 0x00
+						0xb4 0x00 0x00 0x00
+						0xb8 0x08 0x08 0x00
+						0x194 0x06 0x06 0x3e8
+						0x19c 0x01 0x01 0x00
+						0x178 0x00 0x00 0x00
+						0xd0 0x68 0x68 0x00
+						0xdc 0xab 0xab 0x00
+						0xe0 0xaa 0xaa 0x00
+						0xe4 0x02 0x02 0x00
+						0x78 0x09 0x09 0x00
+						0x84 0x16 0x16 0x00
+						0x90 0x28 0x28 0x00
+						0x108 0xa0 0xa0 0x00
+						0x10c 0x00 0x00 0x00
+						0x184 0x0a 0x0a 0x00
+						0x4c 0xaa 0xaa 0x00
+						0x50 0x29 0x29 0x00
+						0x54 0x00 0x00 0x00
+						0xc8 0x00 0x00 0x00
+						0x18c 0x00 0x00 0x00
+						0xcc 0x00 0x00 0x00
+						0x128 0x00 0x00 0x00
+						0x0c 0x0a 0x0a 0x00
+						0x10 0x01 0x01 0x00
+						0x1c 0x7d 0x7d 0x00
+						0x20 0x01 0x01 0x00
+						0x14 0x00 0x00 0x00
+						0x18 0x00 0x00 0x00
+						0x24 0x0a 0x0a 0x00
+						0x28 0x05 0x05 0x00
+						0x48 0x0f 0x0f 0x00
+						0x70 0x0f 0x0f 0x00
+						0x100 0x80 0x80 0x00
+						0x440 0x0b 0x0b 0x00
+						0x4d8 0x02 0x02 0x00
+						0x4dc 0x6c 0x6c 0x00
+						0x4e0 0xbb 0xb8 0x00
+						0x508 0x77 0x77 0x00
+						0x50c 0x80 0x80 0x00
+						0x514 0x03 0x03 0x00
+						0x51c 0x16 0x16 0x00
+						0x448 0x75 0x75 0x00
+						0x454 0x00 0x00 0x00
+						0x40c 0x0a 0x0a 0x00
+						0x41c 0x06 0x06 0x00
+						0x510 0x00 0x00 0x00
+						0x268 0x45 0x45 0x00
+						0x2ac 0x12 0x12 0x00
+						0x294 0x06 0x06 0x00
+						0x254 0x00 0x00 0x00
+						0x8c8 0x83 0x83 0x00
+						0x8c4 0x02 0x02 0x00
+						0x8cc 0x09 0x09 0x00
+						0x8d0 0xa2 0xa2 0x00
+						0x8d4 0x85 0x85 0x00
+						0x880 0xd1 0xd1 0x00
+						0x884 0x1f 0x1f 0x00
+						0x888 0x47 0x47 0x00
+						0x80c 0x9f 0x9f 0x00
+						0x824 0x17 0x17 0x00
+						0x828 0x0f 0x0f 0x00
+						0x8b8 0x75 0x75 0x00
+						0x8bc 0x13 0x13 0x00
+						0x8b0 0x86 0x86 0x00
+						0x8a0 0x04 0x04 0x00
+						0x88c 0x44 0x44 0x00
+						0x870 0xe7 0xe7 0x00
+						0x874 0x03 0x03 0x00
+						0x878 0x40 0x40 0x00
+						0x87c 0x00 0x00 0x00
+						0x9d8 0x88 0x88 0x00
+						0xffffffff 0xffffffff 0x00 0x00>;
+			qcom,qmp-phy-reg-offset = <0x988 0x98c 0x990 0x994
+						0x974 0x8d8 0x8dc 0x804 0x800
+						0x808>;
+
+			clocks = <&gcc GCC_USB0_AUX_CLK>,
+				 <&gcc GCC_USB0_PIPE_CLK>;
+
+			clock-names = "aux_clk", "pipe_clk";
+
+			resets =  <&gcc GCC_USB0_PHY_BCR>,
+				  <&gcc GCC_USB3PHY_0_PHY_BCR>;
+			reset-names = "usb3_phy_reset",
+				      "usb3phy_phy_reset";
+
+			status = "disabled";
+		 };
+
+		qusb_phy_0: qusb@79000 {
+			compatible = "qcom,qusb2phy";
+			reg = <0x079000 0x180>,
+			      <0x08af8800 0x400>,
+			      <0x01841030 0x4>,
+			      <0x08A0C12C 0x4>;
+			reg-names = "qusb_phy_base",
+				    "qscratch_base",
+				    "ref_clk_addr",
+				    "usb3_guctl_addr";
+
+			qcom,qusb-phy-init-seq = <0x14 0x00
+						  0xF8 0x80
+						  0xB3 0x84
+						  0x83 0x88
+						  0xC0 0x8C
+						  0x30 0x08
+						  0x79 0x0C
+						  0x21 0x10
+						  0x00 0x90
+						  0x00 0x18
+						  0x14 0x9C
+						  0x80 0x04
+						  0x9F 0x1C>;
+			phy_type= "utmi";
+
+			resets = <&gcc GCC_QUSB2_0_PHY_BCR>;
+			reset-names = "usb2_phy_reset";
+
+			status = "disabled";
+		};
+
+		dbm_1p5: dbm@0x8AF8000 {
+			compatible = "qcom,usb-dbm-1p5";
+			reg = <0x8AF8000 0x300>;
+			qcom,reset-ep-after-lpm-resume;
+		};
+
+		usb3: usb3@8A00000 {
+			compatible = "qcom,ipq6018-dwc3";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			ranges;
+			reg = <0x8AF8800 0x100>,
+				<0x8A00000 0xcd00>;
+			reg-names = "qscratch_base", "dwc3_base";
+			clocks = <&gcc GCC_SNOC_BUS_TIMEOUT2_AHB_CLK>,
+				 <&gcc GCC_SYS_NOC_USB0_AXI_CLK>,
+				 <&gcc GCC_USB0_MASTER_CLK>,
+				 <&gcc GCC_USB0_SLEEP_CLK>,
+				 <&gcc GCC_USB0_MOCK_UTMI_CLK>,
+				 <&gcc GCC_USB0_PHY_CFG_AHB_CLK>,
+				 <&gcc GCC_USB0_AUX_CLK>,
+				 <&gcc GCC_USB0_PIPE_CLK>;
+			clock-names = "snoc_bus_timeout2",
+				      "sys_noc_axi",
+				      "master",
+				      "sleep",
+				      "mock_utmi",
+				      "cfg_ahb_clk",
+				      "aux_clk",
+				      "pipe_clk";
+			assigned-clocks = <&gcc GCC_SYS_NOC_USB0_AXI_CLK>,
+					  <&gcc GCC_USB0_MASTER_CLK>,
+					  <&gcc GCC_USB0_MOCK_UTMI_CLK>;
+			assigned-clock-rates = <133330000>,
+						<133330000>,
+						<24000000>;
+			qca,host = <1>;
+			qcom,usb-dbm = <&dbm_1p5>;
+			status = "disabled";
+
+			dwc_0: dwc3@8A00000 {
+			       compatible = "snps,dwc3";
+			       reg = <0x8A00000 0xcd00>;
+			       interrupts = <GIC_SPI 140 IRQ_TYPE_NONE>;
+			       usb-phy = <&qusb_phy_0>, <&ssphy_0>;
+			       snps,dis_ep_cache_eviction;
+			       tx-fifo-resize;
+			       snps,usb3-u1u2-disable;
+			       snps,nominal-elastic-buffer;
+			       snps,is-utmi-l1-suspend;
+			       snps,hird-threshold = /bits/ 8 <0x0>;
+			       snps,dis_u2_susphy_quirk;
+			       snps,dis_u3_susphy_quirk;
+				   snps,quirk-ref-clock-adjustment = <0xA87F0>;
+				   snps,quirk-ref-clock-period = <0x29>;
+			       dr_mode = "host";
+			};
+		};
+
+		qcom,usbbam@8B04000 {
+			compatible = "qcom,usb-bam-msm";
+			reg = <0x8B04000 0x17000>;
+			interrupt-parent = <&intc>;
+			interrupts = <GIC_SPI 135 IRQ_TYPE_NONE>;
+
+			qcom,bam-type = <0>;
+			qcom,usb-bam-fifo-baseaddr = <0x4A100000>;
+			qcom,usb-bam-num-pipes = <4>;
+			qcom,ignore-core-reset-ack;
+			qcom,disable-clk-gating;
+			qcom,usb-bam-override-threshold = <0x4001>;
+			qcom,usb-bam-max-mbps-highspeed = <400>;
+			qcom,usb-bam-max-mbps-superspeed = <3600>;
+			qcom,reset-bam-on-connect;
+
+			qcom,pipe0 {
+				label = "ssusb-qdss-in-0";
+				qcom,usb-bam-mem-type = <2>;
+				qcom,dir = <1>;
+				qcom,pipe-num = <0>;
+				qcom,peer-bam = <0>;
+				qcom,peer-bam-physical-address = <0x6064000>;
+				qcom,src-bam-pipe-index = <0>;
+				qcom,dst-bam-pipe-index = <0>;
+				qcom,data-fifo-offset = <0x0>;
+				qcom,data-fifo-size = <0xe00>;
+				qcom,descriptor-fifo-offset = <0xe00>;
+				qcom,descriptor-fifo-size = <0x200>;
+			};
+		};
+
+		extcon_usb: extcon_usb {
+			compatible = "linux,extcon-usb-gpio";
+			status = "disabled";
+		};
+
 		qusb_phy_1: qusb@59000 {
 			compatible = "qcom,ipq6018-qusb2-phy";
 			reg = <0x0 0x059000 0x0 0x180>;
